File: src/chipschallenge/blockreactions/FirebootsTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Inventory.Boots;

/**
 * Move to fire boots
 */
public class FirebootsTo extends BlockReaction {

    private FirebootsTo() {
    }
    private static FirebootsTo mInstance = null;

    public static synchronized FirebootsTo getInstance() {
        if (mInstance == null) {
            mInstance = new FirebootsTo();
        }
        return mInstance;
    }

    // Chip can pick fireboots
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        if (moving.isChip()) {
            takeBoots(Boots.FIREBOOTS);
            standing.replace(createBlock(Type.FLOOR));
        }
    }

    // Only chip and blocks
    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() || moving.isBlock();
    }
}

File: ChipsChallenge.Shared/Blockreactions/FirebootsTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Boots = Inventory.Boots;

    public class FirebootsTo : NoSlipReaction
    {
        private FirebootsTo()
        {
        }
        private static FirebootsTo instance;

        public static FirebootsTo Instance
        {
            get
            {
                lock (typeof(FirebootsTo))
                {
                    return instance ?? (instance = new FirebootsTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeBoots(Boots.FIREBOOTS);
                standing.Replace(CreateBlock(Block.Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip || moving.IsBlock();
        }
    }
}